Add Unstoppable Wallet

esengulov opened the pull request:
Fully decentralized SPV wallet for Bitcoin and other crypto. Well designed and built fully using native technologies with best security practices in mind.

- https://unstoppable.money
- source code iOS: https://github.com/horizontalsystems/unstoppable-wallet-ios
- source code Android: https://github.com/horizontalsystems/unstoppable-wallet-android

crwatkins commented:
@esengulov Thanks for the submission!

Note that in the comment above you provide the iOS repository for Android (not an issue).

Regretfully, there are a few PRs currently in the queue, so the review will not start immediately.
You can help speed up the process by making sure that the wallet strictly complies with all of the [listing criteria](https://github.com/bitcoin-dot-org/bitcoin.org/blob/master/docs/managing-wallets.md) and requesting any clarifications or discussing any discrepancies here in the meantime.

After one quick look, one optional criterion you might want to consider is

* Does not show "received from" Bitcoin addresses in the UI

How many confirmations are required before a transaction will show up in the balance?  I have seen four confirmations, but no balance.

What BIP32 path do you use for Bitcoin addresses?

esengulov commented:
> @esengulov Thanks for the submission!
> 
> Note that in the comment above you provide the iOS repository for Android (not an issue).
> 

my bad, corrected that link!

> Regretfully, there are a few PRs currently in the queue, so the review will not start immediately.
> You can help speed up the process by making sure that the wallet strictly complies with all of the [listing criteria](https://github.com/bitcoin-dot-org/bitcoin.org/blob/master/docs/managing-wallets.md) and requesting any clarifications or discussing any discrepancies here in the meantime.

Based on the criteria listed I think we are good on all points and somewhat behind on the items listed below:

> Received independent security audit(s)

We would love to have more developers review and subsequently reuse our code but this is something we have little control over. We keep our entire work public not just by releasing our repositories but also keeping the entire development process public, down to the every single commit on a regular basis. As more people fork/reuse our libraries the more exposure/audits the codebase will get.

It goes without saying that while building the wallet we started by analyzing the currently available libraries and made sure that libraries we release (bitcoin-spv-kit-ios, bitcoin-spv-kit-android) abide by all security standards. At this particular point we believe there is no wallet that's being developed and improved as throughly as the unstoppable wallet from an engineering perspective. About 10 senior level full stack engineers been working on the wallet and its underlying infrastructure for about a year now. 

> Does not show "received from" Bitcoin addresses in the UI

We deliberately added from address to the UI as from privacy perspective this has no advantages/disadvantages as such. In other words, when I receive a Bitcoin payment from someone being able to see an address where the payment came from doesn't really make sender's identity any more private. This is mainly because every transaction is publicly available and can be easily looked up on any block explorer. Surely, we can hide the transaction's "from address" in UI but then we are also effectively blocking the user from cross checking any pending/completed transactions that appear in the UI (Transactions tab) against the publicly available Bitcoin block explorers. 

At the moment the wallet allows to cross check TX details that appear in the UI against following third party block explorers:

- BTC.com
- blockchair.com
- horizontalsystems.xyz

Demonstrating the flow using screenshots below:

Here is the Transactions tab, displaying the transactions list:

![IMG_3075](https://user-images.githubusercontent.com/8611762/61931743-57dad500-afa3-11e9-98ef-e90797f50e91.PNG)

Let's look at incoming transaction 0.8695 btc (8,421.31 USD) on June 21st. Clicking on that  transaction the wallet shows following info

![IMG_3080](https://user-images.githubusercontent.com/8611762/61931941-ee0efb00-afa3-11e9-831f-a48395855a9a.PNG)

This info (with the exception of the historical rate) is obtained from corresponding transaction directly from blockchain itself, without relying on a third party middle men (i.e block explorer). 

 The Verify button enables user to independently cross check that data (and lookup further details about transaction that's not possible to extract from blockchain in an SPV manner i.e. transaction fee rate) against other independent public block explorers such as BTC.com. 

![IMG_3081](https://user-images.githubusercontent.com/8611762/61932296-d421e800-afa4-11e9-8912-2238e52f90ab.PNG)

This page pulls the data from third party public block explorers without any modifications. On that page, the source cell enables user to further cross check the presented info against several publicly available block explorers

![IMG_3082](https://user-images.githubusercontent.com/8611762/61932326-ea2fa880-afa4-11e9-8579-131f29dc61bb.PNG)

Long story short, we can remove the FROM field in transaction details but that not serve any purpose other than restrict user's ability to easily send back or refund payments. May not be a strong argument as such but there are no downsides either :) I am open to hear your ideas and motivation behind hiding that field. We don't have strong opinion for/against showing the from address.

> Provides a bug reporting policy on the website

We do provide emails on the App Store and Play Market pages which is being maintained and acted upon when such emails are received. On top we are planning to add that to settings page on the wallet itself.

> How many confirmations are required before a transaction will show up in the balance? I have seen four confirmations, but no balance.

- for the outgoing transactions: the balance gets subtracted as soon as the payments hits the mempool

- for the incoming transactions: For the Bitcoin blockchain we go with the recommended 6 blocks before the transaction reflected on the balance tab and can be used as an input in the transactions. For other blockchains it varies based on the specifics of that particular blockchain.

> What BIP32 path do you use for Bitcoin addresses?

We use BIP44 which fully supports BIP32 as well as ability to have multi-accounts, https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki

Let me know if you have any other questions

esengulov commented:
Follow up comment on FROM address:

I just had a quick look over at https://en.bitcoin.it/wiki/From_address and realized that there are some valid concerns we haven't thought through. We decided to remove the "FROM" field from the TX info controller.

- https://github.com/horizontalsystems/unstoppable-wallet-android/issues/1061
- https://github.com/horizontalsystems/unstoppable-wallet-ios/issues/790

crwatkins commented:
@esengulov 
> We decided to remove the "FROM" field from the TX info controller.

Bravo!

> We use BIP44 ...

Are you using BIP141 serialization?  If so, is there a reason you are not also using a BIP49 path?

esengulov commented:
> Are you using BIP141 serialization? If so, is there a reason you are not also using a BIP49 path?

Good question!

We do use serialization scheme shown in BIP 141. However, at the moment, we have chosen to use BIP44 path for derivation to enable smooth transition from older/popular wallets (which mostly support BIP44 at this point i.e. Blockchain.info/com).

crwatkins commented:
> We do use serialization scheme shown in BIP 141. However, at the moment, we have chosen to use BIP44 path for derivation to enable smooth transition from older/popular wallets (which mostly support BIP44 at this point i.e. Blockchain.info/com).

The strangest part is that while I'm observing BIP141, P2WPKH-P2SH, with a BIP44 path, when sending transactions, you seem to be using P2PKH (with BIP44) change addresses.  Is that correct?

crwatkins commented:
@esengulov Just in case you are wondering why I care or where I'm going with this, about 25% of the wallets listed on bitcoin.org just a few years ago are no longer being supported and as much as I wish you continued success, users need to be prepared for apps going away because of changes in development, app stores, operating system upgrades, lost devices, etc.  Some of your users may need to restore their BIP39 phrase in the future (perhaps in a state of panic) and restoring a wallet that receives using p2wpkh-p2sh addresses on m/44' and sends change to p2pkh addresses would be a daunting task for even fairly technical users.

esengulov commented:
> > We do use serialization scheme shown in BIP 141. However, at the moment, we have chosen to use BIP44 path for derivation to enable smooth transition from older/popular wallets (which mostly support BIP44 at this point i.e. Blockchain.info/com).
> 
> The strangest part is that while I'm observing BIP141, P2WPKH-P2SH, with a BIP44 path, when sending transactions, you seem to be using P2PKH (with BIP44) change addresses. Is that correct?

Indeed. Had a check with devs, looks like we left that part for later. Thought we had an active issue for that in place in kit Libraries but seems not. We’ll have it addressed before next release schedule for end of next week.

esengulov commented:
> @esengulov Just in case you are wondering why I care or where I'm going with this, about 25% of the wallets listed on bitcoin.org just a few years ago are no longer being supported and as much as I wish you continued success, users need to be prepared for apps going away because of changes in development, app stores, operating system upgrades, lost devices, etc. Some of your users may need to restore their BIP39 phrase in the future (perhaps in a state of panic) and restoring a wallet that receives using p2wpkh-p2sh addresses on m/44' and sends change to p2pkh addresses would be a daunting task for even fairly technical users.

We implemented  witness support  fairly recently, some parts of it like the change address were left out. We have to address a lot of areas at once...   We’ll patch that to stay consistent in paths 🙂

esengulov commented:
@crwatkins hey Craig, we have addressed all mentioned concerns, the last issue should be closed by the time you read this :)

iOS:
- https://github.com/horizontalsystems/bitcoin-kit-ios/issues/407
- https://github.com/horizontalsystems/unstoppable-wallet-ios/issues/790

Android:
- https://github.com/horizontalsystems/bitcoin-kit-android/issues/399
- https://github.com/horizontalsystems/unstoppable-wallet-android/issues/1061

wbnns commented on _wallets/unstoppable.md:
```diff
@@ -0,0 +1,42 @@
+    default: &DEFAULT
+      text: "walletunstoppable"
+      link: "https://unstoppable.money"
+      source: "https://github.com/horizontalsystems"
```
This should be a direct link to the repository containing the wallet's source code.

esengulov replied:
> This should be a direct link to the repository containing the wallet's source code.

The reason I linked this to organization page because there are two repositories for the app, one for iOS, one for Android... Both are pinned on org page.

Can change the link to point to one of them if you prefer that

wbnns replied:
Yes, please - [here's an example](https://github.com/bitcoin-dot-org/bitcoin.org/blob/master/_wallets/bither.md) of how to link to multiple repositories.

esengulov replied:
> Yes, please - [here's an example](https://github.com/bitcoin-dot-org/bitcoin.org/blob/master/_wallets/bither.md) of how to link to multiple repositories.

thanks, done!

wbnns commented on _translations/en.yml:
```diff
@@ -207,6 +207,7 @@ en:
     walletblw: "Mobile wallet with Lightning Network support."
     walleteclairmobile: "Eclair Mobile is a next generation, Lightning-ready Bitcoin wallet. It can be used as a regular Bitcoin wallet, and can also connect to the Lightning Network for cheap and instant payments."
     walletwasabi: "Wasabi is an open-source, non-custodial, privacy focused Bitcoin wallet for Windows, Linux and Mac. Built-in Tor anonymity network, CoinJoin and coin control features."
+    walletunstoppable: "Unstoppable is one of the easiest and most advanced wallets for Bitcoin. It's built with best security and usability practices in mind. It is designed to communicate to the Bitcoin network directly without intermediaries."
```
Please do not use superlatives or exclusive phrases in the wallet description, such as:

+ "one of the easiest and most advanced wallets"
+ "best security and usability practices"

crwatkins replied:
We've added more guidance on creating the description [here](https://github.com/bitcoin-dot-org/bitcoin.org/blob/master/docs/managing-wallets.md#adding-a-wallet) since you contributed your listing (yes, just days ago; no new requirements, just a clarification).

esengulov replied:
> We've added more guidance on creating the description [here](https://github.com/bitcoin-dot-org/bitcoin.org/blob/master/docs/managing-wallets.md#adding-a-wallet) since you contributed your listing (yes, just days ago; no new requirements, just a clarification).

got it! agree on not using superlatives, was in a hurry when submitting it for the first time ;)

crwatkins replied:
Very good now, however you might be missing the word "a" in the English description.  Thanks.

esengulov replied:
> Very good now, however you might be missing the word "a" in the English description. Thanks.

thanks) done

crwatkins commented on _wallets/unstoppable.md:
```diff
@@ -0,0 +1,46 @@
+          privacynetwork: "checkfailprivacynetworknosupporttor"
+      - name: ios
+        text: "walletunstoppable"
+        link: "https://itunes.apple.com/app/bank-bitcoin-wallet/id1447619907?ls=1&mt=8"
```
Coin also currently uses ?mt=8 so that part is probably not tripping our syntax checker.  None of our other entries use ls=1 so perhaps it is the &?  Those parameters tell iOS what type of media it is (8 is an iOS app) so that they know what app to open (and perhaps skip the web page in between).  The other parameter is also related to skipping the web page.

esengulov commented on _translations/de.yml:
```diff
@@ -199,6 +199,7 @@ de:
     walletbtccom: "Sicher und einfach zu verwenden, bietet die BTC.com Bitcoin-Wallet Ihnen, dem Nutzer, die volle Kontrolle über Ihre Bitcoins. Features beinhalten HD- und Multisignatur-Technologie, Multi-Plattform-Unterstützung für verschiedene Sprachen und Währungen, QR Codes und mehr."
     walleteclairmobile: "Eclair Mobile ist eine Lightning-fähige Bitcoin-Wallet der nächsten Generation. Sie kann wie eine normale Wallet verwendet werden, sich aber für günstige und schnelle Zahlungen mit dem Lightning-Netzwerk verbinden."
     walletwasabi: "Wasabi ist eine ##non-custodial## Open Source-Bitcoin-Wallet. Sie legt den Fokus auf Privatsphäre und läuft unter Windows, Linux und Mac. Sie integriert das Tor-Netzwerk, CoinJoin und Features zur Kontrolle der Coins."
+    walletunstoppable: "Die Unstoppable-Wallet ist eine sichere und vollständig dezentralisierte Kryptobörse, die die wichtigsten Kryptowährungen einhält, darunter Bitcoin(BTC), Ethereum(ETH), Dash(DASH), Bitcoin Cash(BCH), ERC 20-Token und andere."
```
> The localized wallet descriptions do not appear to match the English description and are more general in nature, describing altcoins and referencing ERC-20 tokens.

good point, will fix them shortly ;)

esengulov commented:
@wbnns @crwatkins 

Update! The version 0.7 for the Unstoppable Wallet is out 

- https://github.com/horizontalsystems/unstoppable-wallet-ios/releases
- https://github.com/horizontalsystems/unstoppable-wallet-android/releases

The wallet got a major update. Apart from the new functionality, all requested changes are now in production. There were a number of minor design related recommendations, some of them are still work in progress.

crwatkins commented:
@esengulov Thanks for the update.  I see that change now goes to P2WPKH but the path is still a BIP44 path.  For efficiency, that's a good idea, but for restore compatibility with other (hardware and software) wallets it could be problematic.  I don't know of any other wallet is going to look for P2WPKH addresses on a BIP44 path.  Is there some particular reason you are using BIP44?

esengulov commented:
> @esengulov Thanks for the update. I see that change now goes to P2WPKH but the path is still a BIP44 path. For efficiency, that's a good idea, but for restore compatibility with other (hardware and software) wallets it could be problematic. I don't know of any other wallet is going to look for P2WPKH addresses on a BIP44 path. Is there some particular reason you are using BIP44?

@crwatkins We kept the BIP44 path primarily to enable easy migration for users from older wallets (i.e. blockchain info) to Unstoppable. Based on the install count data shown on Play Market (https://play.google.com/store/apps/details?id=piuk.blockchain.android), it does seem to be as the most popular wallet around, by a wide margin. So we wanted to enable those users to migrate to a contemporary wallet. 

That said, your point regarding potential migration options from Unstoppable to other wallets indeed is a valid concern... At this point, it might be problematic for those users to find the wallet that would retrieve all addresses correctly and that's a problem. I am going to work on this next few days and report back on Monday. I am guessing we can upgrade the derivation path to BIP49 by default and enable "BIP44 mode" for those who need it... Will try to have it addressed by the end of the next week.

crwatkins commented:
@esengulov I personally think that sounds like a good idea, but just to be clear, there are no listing criteria which specify any requirements about derivation paths.  Even the criterion that recommends BIP32 HD is optional.

esengulov commented:
> @esengulov I personally think that sounds like a good idea, but just to be clear, there are no listing criteria which specify any requirements about derivation paths. Even the criterion that recommends BIP32 HD is optional.

Thanks for clarifying that Craig :)  I am primarily motivated by providing future and existing users with the most flexible derivation format(s). Once we do that I think the listing will come naturally ;) I am aiming to provide the solution for this tomorrow along with the ready issue to have it implemented next week itself. 

We have made the list of the most widely used wallets over the last 3-4 years and analyzing them over the weekend. One option that seems like a good solution at this point is to have BIP49 compliant derivation paths for all users which setup wallet as a NEW wallet. For users that prefer to migrate existing wallet we can provide a list of wallets they are migrating from. The Unstoppable wallet will continue to operate in the same manner as the original wallet that was used to setup the keys in the first place. 

Work in progress, I will try to followup in detail tomorrow

crwatkins commented:
@esengulov Sounds reasonable.

esengulov commented:
> We have made the list of the most widely used wallets over the last 3-4 years and analyzing them over the weekend. One option that seems like a good solution at this point is to have BIP49 compliant derivation paths for all users which setup wallet as a NEW wallet. For users that prefer to migrate existing wallet we can provide a list of wallets they are migrating from. The Unstoppable wallet will continue to operate in the same manner as the original wallet that was used to setup the keys in the first place.
> 
> Work in progress, I will try to followup in detail tomorrow

So, it took me some time to get to the bottom of this problem and figure out how to proceed. Good news we came up with a good solution that would not involve anything fancy.  

Within the next few weeks we are going to release an update which will enable wallet to operate fully in compliance with either BIP44 or BIP49. 

1. EXISTING users which update from older version of the app to 0.8 or above will start to operate in fully BIP44 compliant manner (no segwit receive). These users will be able to restore without issues using the existing wallet restore option. 

2. NEW users installing version 0.9 or above will have the wallet operating in fully BIP49 compliant manner.

3. MIGRATING users from external BIP44 / BIP49 compliant wallet will continue to operate according to the same BIP.

Will keep you updated!

crwatkins commented:
@esengulov Looks reasonable.

wbnns commented on _translations/tr.yml:
```diff
@@ -118,6 +118,7 @@ tr:
     wallettrezor: "TREZOR kullanım kolaylığından ödün vermeden yüksek seviye güvenlik sağlayan bir cüzdan sistemidir. Soğuk depodan farklı olarak TREZOR işlemleri online bir cihaza bağlıykende imzalayabilir. Bu da güvenli olmayan bir bilgisayar kullanırken bile bitcoin harcamalarınızda güvenlik sağlar."
     walletnano: "Ledger Nano, ST23YT66 bankacılık akıllı kart platformu üzerine kurulu bir cüzdandır. Kullanıcının özel anahtarlarını güvende tutar, işlemleri onaylar, güvenli bir ön-ödemeli kart olarak yada çoklu-imza partisi olarak kullanılabilir. Açık-kaynak kodlu olmasa da belirleyici şekilde onaylanabilir olabilir."
     walletwasabi: "Wasabi gizlilik odaklı bir bitcoin cüzdanı."
+    walletunstoppable: "Unstoppable Cüzdanı - Bitcoin(BTC), Ethereum(ETH) ve EOS(EOS) gibi büyük kripto para birimlerini destekleyen guzel bir şifreleme cüzdanı uygulamasıdır."
```
Please ensure that the description is consistent with the English description.

wbnns commented on _translations/fr.yml:
```diff
@@ -121,6 +121,7 @@ fr:
     walletninki: "Ninki est un portefeuille multi-signature avec une interface utilisateur magnifique qui rend l'envoi de bitcoins aussi facile qu'un email. Vous avez le contrôle total de vos bitcoins en tout temps."
     walleteclairmobile: "Eclair Mobile est un portefeuille Bitcoin de nouvelle génération, prêt pour Lightning. Il peut s'utiliser comme un portefeuille Bitcoin standard, et peut aussi se connecter au réseau Lightning, pour des paiements rapides et instantanés."
     walletwasabi: "Wasabi est un portefeuille Bitcoin protégeant votre vie privée."
+    walletunstoppable: "Unstoppable est un portefeuille crypté sécurisé et entièrement décentralisé prenant en charge des principales crypto-monnaies, notamment les jetons Bitcoin."
```
Please ensure that the description is consistent with the English description.

esengulov replied:
done

wbnns commented on _translations/ko.yml:
```diff
@@ -118,6 +118,7 @@ ko:
     wallettrezor: "TREZOR는 편리함을 희생하지 않고 높은 수준의 보안을 제공하는 하드웨어 지갑입니다. 콜드 스토리지와 달리, TREZOR는 온라인 장치에 연결된 상태에서 거래에 서명할 수 있습니다. 즉, 비트코인을 사용하는 것은 위험에 노출된 컴퓨터를 사용할 시에도 안전하다는 것을 의미합니다."
     walletnano: "Ledger Nano는 하드웨어 지갑이며 ST23YT66 은행 스마트카드 플랫폼으로 만들어졌습니다. 사용자 개인키를 안전하게 보호하고, 보안 선불 카드 또는 멀티 서명으로 사용되는 거래를 확인합니다. 오픈 소스가 아닌 경우, 엄격하게 확인할 수 있습니다."
     walletwasabi: "와사비는 개인정보 보호를 중요시 여기는 비트코인 지갑입니다."
+    walletunstoppable: "월렛 뱅크는 비트코인, 이더리움, 비트코인 케쉬, 이알씨 20 토크들들을 포함한 주요한 암호통화를 지킬 안전하고 완전하게 분산화된 비밀 지갑 어플입니다. 지원되는 통화들의 수는 점점 추가되어 배포할 것입니다."
```
Please ensure that the description is consistent with the English description.

esengulov replied:
done

wbnns commented on _translations/ru.yml:
```diff
@@ -176,6 +176,7 @@ ru:
     walletbtccom: "Безопасный и простой в использовании, BTC.com Bitcoin Wallet позволяет вам, пользователю, полностью контролировать ваши биткойны. Особенности включают технологию HD и мульти-подпись, многоплатформенную, местную языковую и валютную поддержку, QR-коды и многое другое."
     walletblw: "Мобильный кошелек с поддержкой Lightning Network."
     walletwasabi: "Некастодиальный, фокусированный на конфиденциальности биткойн-кошелек с открытым исходным кодом, для Windows, Linux и Mac. Встроенные анонимная сеть Tor, CoinJoin и функции контроля койнов."
+    walletunstoppable: "Unstoppable - это безопасный, красивый и полностью децентрализованный крипто-кошелек, который поддерживает Bitcoin (BTC) и другие криптовалюты"
```
Please ensure that the description is consistent with the English description.

esengulov commented:
> @esengulov
> 
> Hello, we left a few additional inline comments with requested changes.
> 
> Also, the features for Unstoppable Wallet need to be added to `_wallets/unstoppable.md ` - please see the **Features** section in the [adding a wallet](https://github.com/bitcoin-dot-org/bitcoin.org/blob/master/docs/managing-wallets.md#adding-a-wallet) area of the managing wallets documentation.
> 
> Please let us know if you have any questions.

done!

crwatkins commented:
I have reviewed the Unstoppable wallet from Horizontal Systems based on the current wallet requirements criteria and my evaluation is below. The summary is that the wallet passes on security and overall design. However, because the wallet is new and has received very little feedback, I cannot recommend it for listing at this time. I will be glad to recommend Unstoppable for listing after the following items are addressed:

1. The wallet has received sufficient user feedback
1. The unstoppable.money web site security (HSTS) has been improved
1. Bug reporting methods have been provided on the web site
1. Some information about the developers has been provided on the web site

All of these items are described in more detail in the review below.  I recommend revisiting this listing in six months.   

I concur with the scoring in 2345bad.

---


## Unstoppable Wallet
### iOS version 0.8
### Android version 0.8.0
### Review Version 2019091201


The wallet list is based on the personal evaluation of the maintainer(s) and
regular contributors of this site, according to the criteria detailed below.

These requirements are meant to be updated and strengthened over time.
Innovative wallets are exciting and encouraged, so if your wallet has a good
reason for not following some of the rules below, please submit it anyway and
we'll consider updating the rules.

**NOTE** The version reviewed requires 6 confirmations before transactions can be spent instead of allowing the user the discretion to decide.  This seems that it would make many payment situations difficult.  Horizontal Systems claims this will be adjusted in the future.

Basic requirements:

- Sufficient users and/or developers feedback can be found without concerning
  issues, or independent security audit(s) is available
  
  **FAIL** Very little feedback on Apple App Store, Google Play, Reddit, Bitcointalk, or by using generic web searches
  
- No indication that users have been harmed considerably by any issue in
  relation to the wallet
  
  **PASS** No indication
  
- No indication that security issues have been concealed, ignored, or not
  addressed correctly in order to prevent new or similar issues from happening
in the future

    **PASS** No indication

- No indication that the wallet uses unstable or unsecure libraries

    **PASS** No indiction.  Uses SecureRandom() and SecRandomCopyBytes() for entropy.

- No indication that changes to the code are not properly tested

    **PASS** No indication.  Contains unit tests.

- Wallet was publicly announced and released since at least 3 months

    **PASS** In app store since at least January 2019

- No concerning bug is found when testing the wallet

    **PASS** No concerning bugs found

- Website supports HTTPS and 301 redirects HTTP requests

    **PASS** http://unstoppable.money redirects to https

- SSL certificate passes [Qualys SSL Labs SSL
  test](https://www.ssllabs.com/ssltest/)

    **PASS** https://unstoppable.money A rating

- Website serving executable code or requiring authentication uses HSTS with a
  max-age of at least 180 days

    **FAIL** No HSTS support
  
- The identity of CEOs and/or developers is public

    **FAIL** Horizontal Systems is a DAO and we currently don't have any provisions for such organizations.
    
    **NOTE** We should consider criteria for DAOs in the future, however Horizontal Systems has informed us that they plan on publishing some more information about their organization in the near future that is likely to meet our existing criteria.

- Avoid address reuse by displaying a new receiving address for each transaction
  in the wallet UI

    **PASS** Uses new receiving address for each transaction

- Uses deterministic ECDSA nonces (RFC 6979)

    **PASS** Transactions from both iOS and Android wallets were duplicated and signed with custom code and the signatures were identical.

- Avoid address reuse by using a new change address for each transaction

    **PASS** Uses a new change address for each transaction

- User has access to private keys
- If private keys or encryption keys are stored online:
  - Refuses weak passwords (short passwords and/or common passwords) used to
    secure access to any funds, or provides an aggressive account lock-out
feature in response to failed login attempts along with a strict account
recovery process.

  **N/A**

- If user has exclusive access over its private keys:
  - Allows backup of the wallet

  **PASS** Allows backup using 12 word BIP39 phrase
  
  - Restoring wallet from backup is working

    **PASS** Wallet was restored succesfully
  
  - Source code is public and kept up to date under version control system

    **PASS** https://github.com/horizontalsystems
      
- If user has no access to some of the private keys in a multi-signature wallet:
  - Provides 2FA authentication feature
  - Reminds the user to enable 2FA by email or in the main UI of the wallet
  - User session is not persistent, or requires authentication for spending
  - Gives control to the user over moving their funds out of the multi-signature
    wallet

   **N/A**

- For hardware wallets:
  - Uses the push model (computer malware cannot sign a transaction without user
    input)
  - Protects the seed against unsigned firmware upgrades
  - Supports importing custom seeds
  - Provides source code and/or detailed specification for blackbox testing if
    using a closed-source Secure Element
    
   **N/A**

Optional criteria (some could become requirements):

- Received independent security audit(s)

  **NOTE** No known audits

- Does not show "received from" Bitcoin addresses in the UI

  **PASS** Does not show received Bitcoin addresses
  
- Provides a bug reporting policy on the website

  **FAIL** No bug reporting policy on the website
  
  **NOTE** Horizontal Systems expects to improve contact information shortly
  
- Website serving executable code or requiring authentication is included in the
  [HSTS preload list](https://hstspreload.org/)
  
  **FAIL** https://unstoppable.money does not support HSTS

- If user has exclusive access over its private keys:
  - Supports HD wallets (BIP32)

      **PASS** Supports BIP32 with BIP44 and BIP49 paths
  
  - Provides users with step to print or write their wallet seed on setup

      **PASS** Does not require backup on setup, but does require backup before receiving funds
  
  - Uses a strong KDF and key stretching for wallet storage and backups

      **PASS** The iOS Keychain is used on iOS and the KeyStore system on Android.

- On desktop platform:
    - Encrypt the wallet by default

   **N/A**

- For hardware wallets:
  - Prevents downgrading the firmware

   **N/A**

esengulov commented:
> I have reviewed the Unstoppable wallet from Horizontal Systems based on the current wallet requirements criteria and my evaluation is below. The summary is that the wallet passes on security and overall design. However, because the wallet is new and has received very little feedback, I cannot recommend it for listing at this time. I will be glad to recommend Unstoppable for listing after the following items are addressed:
> 
> 1. The wallet has received sufficient user feedback
> 2. The unstoppable.money web site security (HSTS) has been improved
> 3. Bug reporting methods have been provided on the web site
> 4. Some information about the developers has been provided on the web site
> 
> All of these items are described in more detail in the review below. I recommend revisiting this listing in six months.
> 
> I concur with the scoring in [2345bad](https://github.com/bitcoin-dot-org/bitcoin.org/commit/2345badfaf8fbd5ef9de6e5325687f71662b999b).

Thanks @crwatkins 

We will work on the items you outlined and followup here once all addressed, within 2 weeks.

Peace!

wbnns commented:
Closing for now; we can reopen when the wallet is ready to be reviewed again.

esengulov commented:
> Closing for now; we can reopen when the wallet is ready to be reviewed again.

sounds good, will followup once all issues been addressed

wbnns commented:
@esengulov 

Ok, thanks.

esengulov commented:
special thanks to @crwatkins for the invaluable feedback we received during the review

esengulov commented:
Hey Craig @crwatkins, hey Will @wbnns 

  It’s been a few months since our original PR and the resulting feedback. During that time we had been working non stop on making the wallet better and safer!

  I am hoping that we can now revisit our PR to add Unstoppable to Bitcoin.org's recommended wallets list. Apart from attempting to address all recommended steps we have added a number of essential features that most wallets don’t have.   

At the time of submission the wallet was on version 0.6 - 0.7, and we now at 0.12 on production with 0.13 being prepared for release in about two weeks.  

- Android https://github.com/horizontalsystems/unstoppable-wallet-android/releases
- iOS : https://github.com/horizontalsystems/unstoppable-wallet-ios/releases  

Both platforms are moving in parallel and it goes without saying that everything that has been done/discussed during that time is available on the Github for those interested. We don’t use any other task managers / trackers other than Github, so everything is there.

Below I am going to outline some of the notable changes to the Unstoppable wallet since our last submission. Please note that the wallet is constant work in progress and there are close to 8-10 engineers working on the Unstoppable app full time. At this point we believe Unstoppable to be the most comprehensive and decentralized wallet out there, especially among the currently available mobile wallets.  

## BITCOIN WALLET IMPROVEMENTS

**Cross Wallet Compatibility** 

The Unstoppable wallet is now fully standard compliant and provides ability to setup/export/import wallets according to BIP44 / BIP49 / BIP84 address formats. 

In other words, the wallet created on Unstoppable can be easily migrated to most of the other wallets and vice versa. The user is free to migrate from Unstoppable wallet to another by restoring the private key on a target wallet. 

**Privacy Improvements**

The wallet now includes BIP 69 support making your transactions / history somewhat more private. You may read more about BIP69 here, https://github.com/bitcoin/bips/blob/master/bip-0069.mediawiki

Moreover full TOR integration is close to completion and will be already released on Android in the next update if there are no issues during internal testing. The iOS is likely to take more time as there is some extra work which hasn't been addressed by anyone on iOS before. More details about missing parts are here, https://github.com/iCepa/Tor.framework/issues/54#issuecomment-589612839 

**Bitcoin Transactions**

The wallet now able to identify transactions which were rejected by network nodes and parse them accordingly. I haven’t seen this implemented properly on any SPV wallet. 

Moreover the wallet can correctly identify transactions which have double spend risk, both for incoming and outgoing transactions. When wallet sends/receives a transaction which conflicts with another transaction in the network mempool the wallet will notify user about it by providing the ID for conflicting transaction. It should be noted that this works both for incoming and outgoing transactions.

So when you send or receive a transaction you can be sure that there are no other transactions in the mempool trying to send the same input. 

 **Experimental Features**

As we get more familiar with Bitcoin protocol we started exploring more ways to program transactions. For power users, there is an 'Experimental' section in wallet Settings which enable users to activate certain features which are not available in the wallet by default. 

One and only such feature at this point is to enable user to lock Bitcoins from spending for up to a year. We use this feature internally to force ourselves to HODL part of our income. We have been testing this feature for the last 4 months using our funds, locking in such way about 0.5-0.7 Bitcoins per month.

This feature is not really advertised to the users and disabled by default but those interested can activate it. Once activated the Send interface will display additional parameters for creating transaction i.e. the lock period for the funds. 

While working on this task we have also released that component as a public library:  

- https://github.com/horizontalsystems/bitcoin-kit-android/tree/master/hodler 
- https://github.com/horizontalsystems/bitcoin-kit-ios/tree/master/Hodler

 
## GENERAL IMPROVEMENTS

**Support for More blockchains**

The wallet now supports EOS, Binance DEX and Litecoin blockchains. The Litecoin support is not production yet and is scheduled for release in the version 0.13 which is due end of next week. 

**Code Architecture Improvements**

A lot of architecture improvements to the code base, interface improvements and bug fixes. We put a great deal of importance on keeping the engineering process and code base flexible. While this may not be noticeable on the surface a lot of engineering resources are being put into this.

Our intent is to encourage developers to use existing unstoppable code base as a building block for future wallets and other blockchain based products which require wallet under the hood.  

**Price Monitoring**

The Unstoppable now allows to monitor price changes, follow coin news and trading volumes over various time periods in a sleek and responsive interface. 

The data pulled from CryptoCompare.com's publicly available API.

**Tech Support**

During the previous review process there was feedback in regard to the lack of obvious ways to contact developers or submit bugs. 

To address this we have added ways to contact us over email as well as option to chat with the entire engineering team over Telegram. 

To help with the debugging of potential issues the app now has a so called App Status section from where user can see the state of each wallet in the app i.e. sync status, connected peers, etc. This data available only to the user and was primarily added to enable user to share with engineers some non private data in case of bugs/issues

**Public Feedback / Transparency** 

Our core Bitcoin iOS and Android libraries now take first spot on Google search for “bitcoin kit swift” and “bitcoin kit kotlin”. It wasn’t the case during our previous submission :)

We have reworked our websites to add more transparency in regards to the people behind the project

  - company website : https://horizontalsystems.io
 - wallet website : https://unstoppable.money 
- team : https://horizontalsystems.io/team 
- team on Github : https://github.com/orgs/horizontalsystems/people
- team at work :) https://www.youtube.com/watch?v=-qjzSUdeId0

That’s all I could think of :)  

peace!

iBek Esengulov

crwatkins commented:
@esengulov Thanks for the update!  You guys have been busy.  Could you look into enabling HSTS (preferably with preload) on horizontalsystems.io and unstoppable money?

esengulov commented:
> @esengulov Thanks for the update! You guys have been busy. Could you look into enabling HSTS (preferably with preload) on horizontalsystems.io and unstoppable.money?

done!

crwatkins commented:
[Previously](https://github.com/bitcoin-dot-org/bitcoin.org/pull/3054#issuecomment-531409514), the open issues were:

> 1. The wallet has received sufficient user feedback
> 2. The unstoppable.money web site security (HSTS) has been improved
> 3. Bug reporting methods have been provided on the web site
> 4. Some information about the developers has been provided on the web site

Items 2, 3, and 4 have indeed been remedied, however, regretfully, there is still "Very little feedback on Apple App Store, Google Play, Reddit, Bitcointalk, or by using generic web searches".  I've recently reported a couple bugs and after those are fixed, I recommend revisiting this listing in six months.
